File: 2023/14/14.scala
import scala.io.Source

object ReflectCode {
  def main(args: Array[String]): Unit = {
    val file = Source.fromFile("./14.txt").getLines().toList

    var reflector = file.map(_.trim).toSeq

    def transpose(matrix: Seq[String]): Seq[String] = matrix.transpose.map(_.mkString)

    // Flip reflector
    reflector = transpose(reflector)

    for (i <- reflector.indices) {
      var groups = Seq[String]()
      // Sort reflector by splitting #
      for (group <- reflector(i).split("#")) {
        val newGroup = group.sorted(Ordering[Char].reverse)
        groups = groups :+ newGroup.mkString
      }
      reflector = reflector.updated(i, groups.mkString("#"))
    }

    // Flip reflector back
    reflector = transpose(reflector)

    var count = 0

    for ((row, i) <- reflector.reverse.zipWithIndex) {
      count += row.count(_ == 'O') * (i + 1)
    }

    println(count)
  }
}

File: 2023/14/14.py
file = open("./14.txt").readlines()

reflector = tuple(line.strip() for line in file)

# (O) - rounded rock
#  #  - flat rock
# . - empty space
# https://en.wikipedia.org/wiki/Transpose

def transpose(matrix):
    return list(map("".join, zip(*matrix)))

# flip reflector
reflector = transpose(reflector)

for i, row in enumerate(reflector):
    groups = []
    # sort reflector by splitting #
    for group in row.split("#"):
        new_group = sorted(tuple(group), reverse=True)
        groups.append("".join(new_group))

    reflector[i] = "#".join(groups)

# flip reflector back
reflector = transpose(reflector)

count = 0

for i, row in enumerate(reversed(reflector)):
    count += row.count("O") * (i + 1)

print(count)